contracts/eosio.system/src/voting.cpp

```diff
@@ -65,6 +90,45 @@ namespace eosiosystem {
 
    }
 
+   void system_contract::regproducer( const name& producer, const eosio::public_key& producer_key, const std::string& url, uint16_t location ) {
+      require_auth( producer );
+      check( url.size() < 512, "url too long" );
+
+      check( producer_key.index() < 2, "currently only K1 and R1 producer keys are supported" );
+      check( !is_null_key( producer_key ), "public key should not be the default value" );
+
+      register_producer( producer, convert_to_block_signing_authority( producer_key ), url, location );
+   }
+
+   void system_contract::regproducer2( const name& producer, const eosio::block_signing_authority& producer_authority, const std::string& url, uint16_t location ) {
+      require_auth( producer );
+      check( url.size() < 512, "url too long" );
+
+      std::set<eosio::public_key> unique_keys;
+      std::visit( [&](auto&& auth ) {
```

Comment:
Instead of putting this code here, can we not add library level validations to `eosio::public_key` type?  Wouldn't most users of that type also be interested in `is_null_key` and `is_satisfiable`

Reply:
Added an `is_valid` member function to `block_signing_authority_v0` in CDT. No need for `is_null_key` anymore.